kyo-core/shared/src/main/scala/kyo/streams.scala

```diff
@@ -15,65 +16,68 @@ object Stream:
-            val handler =
-                new ResultHandler[Int, Const[V], Streams[V], Id, Streams[V]]:
-                    def done[T](st: Int, v: T) = v
-                    def resume[T, U: Flat, S2](
-                        st: Int,
-                        command: V,
-                        k: T => U < (Streams[V] & S2)
-                    ) =
-                        if st == 0 then
-                            Resume(0, k(().asInstanceOf[T]))
-                        else
-                            Streams.emitValueAndThen(command)(
-                                Resume(st - 1, k(().asInstanceOf[T]))
-                            )
-            Streams[V].handle(handler)(n, s)
+            if n <= 0 then
+                runDiscard
+            else
+                val handler =
+                    new ResultHandler[Int, Const[Chunk[V]], Streams[V], Id, Streams[V]]:
+                        def done[T](st: Int, v: T) = v
+                        def resume[T, U: Flat, S2](
+                            st: Int,
+                            command: Chunk[V],
+                            k: T => U < (Streams[V] & S2)
+                        ) =
+                            if st == 0 then
+                                Resume(0, k(().asInstanceOf[T]))
+                            else
+                                val t: Chunk[V] = command.take(st)
+                                Streams.emitChunkAndThen(t) {
+                                    Resume(st - t.size, k(().asInstanceOf[T]))
+                                }
+                Streams[V].handle(handler)(n, s)
         end take
 
         def drop(n: Int): Stream[T, V, S] =
-            val handler =
-                new ResultHandler[Int, Const[V], Streams[V], Id, Streams[V]]:
-                    def done[T](st: Int, v: T) = v
-                    def resume[T, U: Flat, S2](
-                        st: Int,
-                        command: V,
-                        k: T => U < (Streams[V] & S2)
-                    ) =
-                        if st == 0 then
-                            Streams.emitValueAndThen(command)(
-                                k(().asInstanceOf[T])
-                            )
-                        else if st == 1 then
-                            k(().asInstanceOf[T])
-                        else
-                            Resume(st - 1, k(().asInstanceOf[T]))
-            Streams[V].handle(handler)(n, s)
+            if n <= 0 then
+                s
+            else
+                val handler =
+                    new ResultHandler[Int, Const[Chunk[V]], Streams[V], Id, Streams[V]]:
+                        def done[T](st: Int, v: T) = v
+                        def resume[T, U: Flat, S2](
+                            st: Int,
+                            command: Chunk[V],
+                            k: T => U < (Streams[V] & S2)
+                        ) =
+                            if st == 0 then
```

Comment:
is it possible for the state to dip below 0?

Reply:
I don't think so. The method checks `n` at the start and then `dropLeft` shouldn't return more elements than `st`.